### DISCUSSION TITLE
Constructing AutoImports with Nuxt kit

### DISCUSSION THREAD

**Discussion by localusercamp** [nuxt/nuxt#16507] [Questions] [answered]
Hi.

I recently discovered `@nuxt/kit` and wanted to experiment with its features. 
I found `addImports`, `addImportsDir` and `addImportsSources`. They are very useful for extracting features from already existing codebases, such as libraries for example. But all of them (as far as I can see) only can import functions from existing file. **_Is there a way to directly pass function to auto-imports?_**

Why do i need this? Imagine that we have a nuxt module that creates auto-import functions depend on files in your project. That module cannot has file inside it's npm package to import newly created functions from it. So is there any way to write thouse functions to some sort of tmp file inside .nuxt folder and then auto-import them from it with `addImports`? Or could we pass source code to be auto-imported?

**Comment by Diizzayy** [comment, marked as answer]
@localusercamp nuxt kit provides a `addTemplate` utility that allows you to write custom code to be saved in the `.nuxt` directory, after doing so, `addImports` can be leveraged to auto import the pertinent functions from the `#build` alias (maps to .nuxt dir).

```ts
addTemplate({
  filename: 'example.mjs',
  getContents: () => [
    'export const useDemo = () => { console.log("demo function") }'
  ].join('\n')
})

addImports({
  from: '#build/example',
  name: 'useDemo'
})

**Reply by localusercamp** [reply]
Freaking awesome! Thank you!

**Reply by localusercamp** [reply]
@Diizzayy Сan these files be typescript, and if so, can it be typed?

**Reply by Diizzayy** [reply]
> Сan these files be typescript, and if so, can it be typed?

@localusercamp Sure, you have two options as it pertains to getting type support. You can either manually creating the type definition file (`example.d.ts`) 

```ts
addTemplate({
  filename: 'example.mjs',
  getContents: () => [
    'export const useDemo = () => { console.log("demo function") }'
  ].join('\n')
})

// manually create the type definition
addTemplate({
  filename: 'example.d.ts',
  getContents: () => [
    'declare const useDemo: () => void;'
  ].join('\n')
})
```


or you can use a typescript file for `addTemplate`:

```ts
// use typescript for the template and have it written to the file directory
addTemplate({
  write: true,
  filename: 'example.ts',
  getContents: () => [
    'export const useDemo = () => { console.log("demo function") }'
  ].join('\n')
})
```

**Reply by localusercamp** [reply]
I created template as typescript file (`example.ts`), but result function was untyped.
It needs to be created with `write: true` as you mentioned. Now all works perfectly.

Thank you for your help!
